sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/healthcare/HL7v2IO.java

```diff
@@ -437,6 +444,20 @@ private Message fetchMessage(HealthcareApiClient client, String msgId)
           .apply(Create.of(this.hl7v2Stores))
           .apply(ParDo.of(new ListHL7v2MessagesFn(this.filter)))
           .setCoder(new HL7v2MessageCoder())
+          // Listing takes a long time for each input element (HL7v2 store) because it has to
+          // paginate through results in a single thread / ProcessElement call in order to keep
+          // track of page token.
+          // Eagerly emit data on 1 second intervals so downstream processing can get started before
+          // all of the list results have been paginated through.
```

Comment:
Unfortunately, this is not possible. If you are paginating from inside the single DoFn `processelement` call, the data coming out of it will only go downstream after the element is done being processed, so this windowing is not changing that in the execution.
This is because bundle execution is committed atomically, so the whole bundle executes before data can go downstream. You do touch on an interesting example, which is one of the reasons that we came up with SplittableDoFn.

Something you could try to do is:
```
PColll<HL7v2Message> pages = hl7v2Stores.apply(ParDo.of(new RetrieveAndOutputPagesFn()))

pages.apply(Reshuffle.viaRandomKey()).apply(ParDo.of(new FetchEachPageFn())
```
Though I don't know if you can actually do that : )

Reply:
Each "page" of responses is a collection of messages. It don't think it make sense to page through all the pages (dropping the real data) to then re-fetch it in the downstream parallelized step. 

In testing w/ customer when pointing at an HL7v2 store with many, many messages (and therefore pages) they reported 
before this change:
there was a long time before any elements were output. so long that they gave up and killed the pipeline. 
after this change: 
there was data coming out more regularly.

This could have been a misunderstanding or a bad test scenario.
I will try to come up with a test that reproduces this behavior.

Reply:
@pabloem does this mean that all of a single element's output must be buffered in memory? or will runner be smart enough to spill to disk?

Based on my initial investigation I was not able to reproduce the behavior reported by customer in a unit test.
summarized in this [gist](https://gist.github.com/jaketf/d3c2e70dde781bbb0ef1993446e34b71)

Reply:
I've opened https://issues.apache.org/jira/browse/BEAM-9856 to explore how this could be done with splittable dofn.